rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+
+        let bytes: Vec<u8> = bits.into_bytes();
+        let secret = BigNum::from_slice(&bytes).unwrap();
+        let mut point = ec::EcPoint::new(&group).unwrap();
+        point.mul_generator(&group, &secret, &mut ctx).unwrap();
+        new_key.set_group(&group).expect("Could not set group");
+        new_key.set_public_key(&point).expect(
+            "Could not set public key",
+        );
+        new_key.set_private_key(&secret).expect(
+            "Could not set private key",
+        );
+        Key(new_key.build())
+    }
+
+    /// Convert the public key into a uncompressed, raw base64 string
+    pub fn to_public_raw(&self) -> String {
+        //Return the public key as a raw bit array
+        let mut ctx = BigNumContext::new().unwrap();
+        let group = Key::group();
+
+        let key = self.0.public_key().unwrap();
+        let keybytes = key.to_bytes(&group, ec::POINT_CONVERSION_UNCOMPRESSED, &mut ctx)
+            .unwrap();
+        base64::encode_config(&keybytes, base64::URL_SAFE_NO_PAD)
+    }
+
+    /// Read the public key from an uncompressed, raw base64 string
+    pub fn from_public_raw(bits: String) -> Result<Key, String> {
+        //Read a public key from a raw bit array
+        let b64bytes = &bits.into_bytes();
+        let bytes: Vec<u8> = base64::decode_config(b64bytes, base64::URL_SAFE_NO_PAD).unwrap();
+        let group = Key::group();
+        if bytes.len() != 65 || bytes[0] != 4 {
+            // It's not a properly tagged key.
+            return Err("Invalid key format for public key.".into());
+        }
+
+        let x = openssl::bn::BigNum::from_slice(&bytes[1..33]).expect("Invalid public key");
+        let y = openssl::bn::BigNum::from_slice(&bytes[33..]).expect("Invalid public key");
+        let mut new_key = match ec::EcKeyBuilder::new() {
+            Ok(key) => key,
+            Err(err) => return Err(format!("Could not generate key: {}", err)),
+        };
+        new_key.set_group(&group).expect("Could not set group");
+        new_key.set_public_key_affine_coordinates(&x, &y).expect(
+            "Invalid coordiates for public key",
+        );
+        Ok(Key(new_key.build()))
+    }
+}
+
+#[derive(Debug)]
+struct AuthElements {
+    t: Vec<String>,
+    k: String,
+}
+
+/// Parse the Authorization Header for useful things.
+fn parse_auth_token(auth_token: &mut String) -> Result<AuthElements, String> {
+    let mut parts: Vec<&str> = auth_token.split(" ").collect();
+    let mut schema = parts.remove(0).to_lowercase();
+    // Ignore the first token if it's the header line.
+    if schema == "authorization:" {
+        schema = parts.remove(0).to_lowercase();
+    }
+    let mut reply: AuthElements = AuthElements {
+        t: Vec::new(),
+        k: String::from(""),
+    };
+    match schema.to_lowercase().as_ref() {
+        "vapid" => {
+            let sub_parts: Vec<&str> = parts[0].splitn(2, ",").collect();
+            for kvi in &sub_parts {
```

Comment:
You can actually avoid the `collect()` and intermediate vector here with something like:

```rust
for kvi in parts[0].splitn(2, ",") {
    // ...
}
```